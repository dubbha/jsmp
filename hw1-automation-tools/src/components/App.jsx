import React, { PropTypes } from 'react';
import Header from './Header';
import ts from '../typescript.ts';
import raw from '../text.txt';

const App = props => (
  <div>
    <Header>Hello, {props.name}!</Header>
    <div>Hey, {props.name}, App Here!</div>
    <div>{ts.greet()}</div>
    <div>{raw}</div>
  </div>
);

App.propTypes = {
  name: PropTypes.string.isRequired
};

export default App;