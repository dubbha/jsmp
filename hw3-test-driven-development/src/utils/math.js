export const add = (a, b) => a + b;

export const sub = (a, b) => a - b;

export const div = (a, b) => (a / b);

export const mult = (a, b) => a * b;

export const pow = (a, b) => (a ** b);

export const double = a => mult(a, 2);

export const triple = a => mult(a, 3);

export const square = a => pow(a, 2);

export const cube = a => pow(a, 3);

export const percent = a => div(a, 100);

export const negate = a => -a;