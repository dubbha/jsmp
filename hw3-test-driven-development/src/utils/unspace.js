const unspace = (str) => {
  if (!str || typeof str !== 'string') return str;

  return str.replace(/\s/g, '');
};

export default unspace;