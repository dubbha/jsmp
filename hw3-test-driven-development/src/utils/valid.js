const valid = (str) => {
  if (!str || typeof str !== 'string') return false;

  const reg = /^\s*\d+\s*(\*\*|[+\-*/])\s*\d+\s*$/;
  if (str.match(reg)) return true;

  return false;
};

export default valid;
