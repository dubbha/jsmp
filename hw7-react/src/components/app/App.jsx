import React from 'react';
import PropTypes from 'prop-types';
import { Route, NavLink } from 'react-router-dom';
import Dashboard from '../dashboard';
import Heroes from '../heroes';
import HeroDetails from '../hero-details';
import './style.sass';
import '../../assets/react.svg';

class App extends React.Component {
  static propTypes = {
    children: PropTypes.element
  };

  constructor(props) {
    super(props);

    this.state = {
      title: 'React Heroes'
    };
  }

  render() {
    return (
      <div className="app">
        <h1 className="app__header">
          <img src="/react.svg" alt="" className="app__logo" />
          {this.state.title}
        </h1>
        <nav>
          <NavLink to="/dashboard" className="app__navLink" activeClassName="active">Dashboard</NavLink>
          <NavLink to="/heroes" className="app__navLink" activeClassName="active">Heroes</NavLink>
        </nav>

        <Route path="/dashboard" component={Dashboard} />
        <Route path="/heroes" component={Heroes} />
        <Route path="/hero-details/:id" component={HeroDetails} />

      </div>
    );
  }
}

export default App;