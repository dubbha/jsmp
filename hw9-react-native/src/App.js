import React from 'react';
import {
  AppRegistry,
  Image,
  StatusBar,
  StyleSheet,
  TouchableOpacity,
  View,
  CameraRoll,
  Alert,
  Modal,
  ScrollView,
  Button,
  TouchableHighlight,
  Dimensions,
} from 'react-native';
import Camera from 'react-native-camera';

import ModalPreview from './components/ModalPreview';
import ModalGallery from './components/ModalGallery';
import Icon from './components/Icon';

const { width } = Dimensions.get('window');

export default class App extends React.Component {
  constructor(props) {
    super(props);

    this.camera = null;

    this.state = {
      camera: {
        aspect: Camera.constants.Aspect.fill,
        captureTarget: Camera.constants.CaptureTarget.cameraRoll,
        type: Camera.constants.Type.back,
        orientation: Camera.constants.Orientation.auto,
        flashMode: Camera.constants.FlashMode.auto,
      },
      photos: [],
      galleryModalVisible: false,
      previewModalVisible: false,
      previewPhoto: '',
    };
  }

  getPhotos = () => {
    CameraRoll.getPhotos({
      first: 9,
      assetType: 'All'
    })
    .then(r => {
      this.setState({ photos: r.edges });
      this.toggleGalleryModal();
    })
  }

  toggleGalleryModal = () => {
    this.setState({ galleryModalVisible: !this.state.galleryModalVisible });
  }

  togglePreviewModal = () => {
    this.setState({ previewModalVisible: !this.state.previewModalVisible });
  }

  takePicture = () => {
    if (this.camera) {
      this.camera.capture()
        .then((data) => console.log(data))
        .catch(err => console.error(err));
    }
  }

  preview = (uri) => {
    this.setState({ previewPhoto: uri });
    this.togglePreviewModal();
  }

  switchType = () => {
    let newType;
    const { back, front } = Camera.constants.Type;

    if (this.state.camera.type === back) {
      newType = front;
    } else if (this.state.camera.type === front) {
      newType = back;
    }

    this.setState({
      camera: {
        ...this.state.camera,
        type: newType,
      },
    });
  }

  get typeIcon() {
    let icon;
    const { back, front } = Camera.constants.Type;

    if (this.state.camera.type === back) {
      icon = require('./assets/ic_camera_rear_white.png');
    } else if (this.state.camera.type === front) {
      icon = require('./assets/ic_camera_front_white.png');
    }

    return icon;
  }

  switchFlash = () => {
    let newFlashMode;
    const { auto, on, off } = Camera.constants.FlashMode;

    if (this.state.camera.flashMode === auto) {
      newFlashMode = on;
    } else if (this.state.camera.flashMode === on) {
      newFlashMode = off;
    } else if (this.state.camera.flashMode === off) {
      newFlashMode = auto;
    }

    this.setState({
      camera: {
        ...this.state.camera,
        flashMode: newFlashMode,
      },
    });
  }

  get flashIcon() {
    let icon;
    const { auto, on, off } = Camera.constants.FlashMode;

    if (this.state.camera.flashMode === auto) {
      icon = require('./assets/ic_flash_auto_white.png');
    } else if (this.state.camera.flashMode === on) {
      icon = require('./assets/ic_flash_on_white.png');
    } else if (this.state.camera.flashMode === off) {
      icon = require('./assets/ic_flash_off_white.png');
    }

    return icon;
  }

  get collectionsIcon() {
    const icon = require('./assets/ic_collections_white.png');
    return icon;
  }

  render() {
    return (
      <View style={styles.container}>
        <StatusBar
          animated
          hidden
        />
        <Camera
          ref={(cam) => {
            this.camera = cam;
          }}
          style={styles.preview}
          aspect={this.state.camera.aspect}
          captureTarget={this.state.camera.captureTarget}
          type={this.state.camera.type}
          flashMode={this.state.camera.flashMode}
          onFocusChanged={() => {}}
          onZoomChanged={() => {}}
          defaultTouchToFocus
          mirrorImage={false}
        />
        <View style={[styles.overlay, styles.topOverlay]}>
          <Icon
            style={styles.typeButton}
            onPress={this.switchType}
            source={this.typeIcon}>
          </Icon>
          <Icon
            style={styles.flashButton}
            onPress={this.switchFlash}
            source={this.flashIcon}>
          </Icon>
          <Icon
            style={styles.collectionsButton}
            onPress={this.getPhotos}
            source={this.collectionsIcon}>
          </Icon>
        </View>
        <View style={[styles.overlay, styles.bottomOverlay]}>
          <TouchableOpacity
              style={styles.captureButton}
              onPress={this.takePicture}
          >
            <Image
                source={require('./assets/ic_photo_camera_36pt.png')}
            />
          </TouchableOpacity>
          <View style={styles.buttonsSpace} />
        </View>

        <ModalGallery
          visible={this.state.galleryModalVisible}
          onToggleGalleryModal={this.toggleGalleryModal}
          photos={this.state.photos}
          onPreview={this.preview}
          styles={styles}
        >
        </ModalGallery>

        <ModalPreview
          visible={this.state.previewModalVisible}
          onTogglePreviewModal={this.togglePreviewModal}
          previewPhoto={this.state.previewPhoto}
          styles={styles}
        >
        </ModalPreview>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  preview: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  overlay: {
    position: 'absolute',
    padding: 16,
    right: 0,
    left: 0,
    alignItems: 'center',
  },
  topOverlay: {
    top: 0,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  bottomOverlay: {
    bottom: 0,
    backgroundColor: 'rgba(0,0,0,0.4)',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  captureButton: {
    padding: 15,
    backgroundColor: 'white',
    borderRadius: 40,
  },
  typeButton: {
    padding: 5,
  },
  flashButton: {
    padding: 5,
  },
  collectionsButton: {
    padding: 5,
  },
  buttonsSpace: {
    width: 10,
  },
  modalContainer: {
    paddingTop: 20,
    flex: 1
  },
  scrollView: {
    flexWrap: 'wrap',
    flexDirection: 'row'
  },
});
