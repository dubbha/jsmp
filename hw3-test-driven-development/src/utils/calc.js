import { add, sub, div, mult, pow } from './math';
import valid from './valid';
import unspace from './unspace';

const calc = (input) => {
  if (!valid(input)) return NaN;

  const unspaced = unspace(input);

  const delim = /(\*\*|[+\-*/])/g;
  const arr = unspaced.split(delim);

  const a = Number.parseInt(arr[0], 10);
  const b = Number.parseInt(arr[2], 10);
  const operator = arr[1];

  switch (operator) {
    case '+':
      return add(a, b);
    case '-':
      return sub(a, b);
    case '/':
      return div(a, b);
    case '*':
      return mult(a, b);
    case '**':
      return pow(a, b);
    default:
      return NaN;
  }
};

export default calc;
