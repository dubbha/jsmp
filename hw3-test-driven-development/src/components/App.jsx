import React from 'react';
import PropTypes from 'prop-types';
import calc from '../utils/calc';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      input: props.input ? props.input : '',
      result: calc(props.input)
    };

    this.calculate = this.calculate.bind(this);
    this.digit = this.digit.bind(this);
    this.oper = this.oper.bind(this);
    this.clear = this.clear.bind(this);
    this.submit = this.submit.bind(this);
  }

  render() {
    return (
      <div>
        <div>
          <form className="calc__form" onSubmit={this.submit} autoComplete="off">
            <input
              className="calc__input"
              type="text"
              id="input"
              value={this.state.input}
              onChange={e => this.setState({ input: e.target.value })} />
            <button
              id="equal"
              className="calc__button calc__button-equal"
              onClick={this.calculate}> = </button>
          </form>
          <span id="result" className="calc__result"> {this.state.result} </span>
        </div>
        <div>
          <button className="calc__button" id="d7" onClick={() => this.digit(7)}> 7 </button>
          <button className="calc__button" id="d8" onClick={() => this.digit(8)}> 8 </button>
          <button className="calc__button" id="d9" onClick={() => this.digit(9)}> 9 </button>
          <button className="calc__button" id="div" onClick={() => this.oper('/')}> &#247; </button>
        </div>
        <div>
          <button className="calc__button" id="d4" onClick={() => this.digit(4)}> 4 </button>
          <button className="calc__button" id="d5" onClick={() => this.digit(5)}> 5 </button>
          <button className="calc__button" id="d6" onClick={() => this.digit(6)}> 6 </button>
          <button className="calc__button" id="mult" onClick={() => this.oper('*')}> &#215; </button>
        </div>
        <div>
          <button className="calc__button" id="d1" onClick={() => this.digit(1)}> 1 </button>
          <button className="calc__button" id="d2" onClick={() => this.digit(2)}> 2 </button>
          <button className="calc__button" id="d3" onClick={() => this.digit(3)}> 3 </button>
          <button className="calc__button" id="add" onClick={() => this.oper('+')}> &#43; </button>
        </div>
        <div>
          <button className="calc__button" id="d0" onClick={() => this.digit(0)}> 0 </button>
          <button
            className="calc__button calc__button-clear"
            id="clear"
            onClick={this.clear}> C </button>
          <button className="calc__button" id="pow" onClick={() => this.oper('**')}> ^ </button>
          <button className="calc__button" id="sub" onClick={() => this.oper('-')}> &#8722; </button>
        </div>
      </div>
    );
  }

  calculate() {
    this.setState(() => ({ result: calc(this.state.input) }));
  }

  digit(num) {
    this.setState({ input: this.state.input + num.toString() });
  }

  oper(str) {
    this.setState({ input: this.state.input + str });
  }

  clear() {
    this.setState({
      input: '',
      result: '',
    });
  }

  submit(e) {
    this.calculate();
    e.preventDefault();
  }
}

App.propTypes = {
  input: PropTypes.string
};

export default App;