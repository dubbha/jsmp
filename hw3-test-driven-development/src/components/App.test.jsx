import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import App from './App';
import * as calc from '../utils/calc';

describe('App', () => {
  beforeEach(() => {
    calc.default = jest.fn((input) => {
      switch (input) {
        case '2+2': return 4;
        case '2+6': return 8;
        case '5*5': return 25;
        case '3**2': return 9;
        case '5**2': return 25;
        case '6**2': return 36;
        case '2**6': return 64;
        case '53-11': return 42;
        case '55/5': return 11;
        case '7*8': return 56;
        default: return null;
      }
    });
  });

  // Testing initial rendering based on props
  // Using react-test-renderer directly, without enzyme
  // Comparing to stored snapshot
  it('should render correctly on startup', () => {
    const comp = renderer.create(<App input="5*5" />).toJSON();
    expect(calc.default).toHaveBeenCalledWith('5*5');
    expect(comp).toMatchSnapshot();
  });

  // Testing DOM manipulation, using enzyme
  it('should re-render correctly on equal button click', () => {
    const comp = shallow(<App input="5**2" />);

    expect(calc.default).toHaveBeenCalledTimes(1);
    expect(calc.default).toHaveBeenCalledWith('5**2');
    expect(comp.find('#result').text().trim()).toBe('25');

    comp.find('#input').simulate('change', { target: { value: '6**2' } });

    expect(calc.default).toHaveBeenCalledTimes(1);
    expect(comp.find('#result').text().trim()).toBe('25');

    comp.find('#equal').simulate('click');

    expect(calc.default).toHaveBeenCalledTimes(2);
    expect(calc.default).toHaveBeenCalledWith('6**2');
    expect(comp.find('#result').text().trim()).toBe('36');
  });

  it('should clear the input on clear button click', () => {
    const comp = shallow(<App input="2+2" />);
    expect(calc.default).toHaveBeenCalledWith('2+2');
    expect(comp.find('#result').text().trim()).toBe('4');

    comp.find('#clear').first().simulate('click');
    expect(comp.find('#result').text().trim()).toBe('');
  });

  it('should enter digits on digit button clicks', () => {
    const comp = shallow(<App />);

    comp.find('#d1').simulate('click');
    comp.find('#d2').simulate('click');
    comp.find('#d3').simulate('click');
    comp.find('#d4').simulate('click');
    comp.find('#d5').simulate('click');
    comp.find('#d6').simulate('click');
    comp.find('#d7').simulate('click');
    comp.find('#d8').simulate('click');
    comp.find('#d9').simulate('click');
    comp.find('#d0').simulate('click');

    expect(comp.state('input')).toBe('1234567890');
  });

  it('should render addition correctly', () => {
    const comp = shallow(<App />);

    comp.find('#d2').simulate('click');
    comp.find('#add').simulate('click');
    comp.find('#d6').simulate('click');

    expect(comp.state('input')).toBe('2+6');

    comp.find('#equal').simulate('click');

    expect(calc.default).toHaveBeenCalledWith('2+6');
    expect(comp.state('result')).toBe(8);
    expect(comp.find('#result').text().trim()).toBe('8');
  });

  it('should render substraction correctly', () => {
    const comp = shallow(<App />);

    comp.find('#d5').simulate('click');
    comp.find('#d3').simulate('click');
    comp.find('#sub').simulate('click');
    comp.find('#d1').simulate('click');
    comp.find('#d1').simulate('click');

    expect(comp.state('input')).toBe('53-11');

    comp.find('#equal').simulate('click');

    expect(calc.default).toHaveBeenCalledWith('53-11');
    expect(comp.state('result')).toBe(42);
    expect(comp.find('#result').text().trim()).toBe('42');
  });

  it('should render division correctly', () => {
    const comp = shallow(<App />);

    comp.find('#d5').simulate('click');
    comp.find('#d5').simulate('click');
    comp.find('#div').simulate('click');
    comp.find('#d5').simulate('click');

    expect(comp.state('input')).toBe('55/5');

    comp.find('#equal').simulate('click');

    expect(calc.default).toHaveBeenCalledWith('55/5');
    expect(comp.state('result')).toBe(11);
    expect(comp.find('#result').text().trim()).toBe('11');
  });

  it('should render multiplication correctly', () => {
    const comp = shallow(<App />);

    comp.find('#d7').simulate('click');
    comp.find('#mult').simulate('click');
    comp.find('#d8').simulate('click');

    expect(comp.state('input')).toBe('7*8');

    comp.find('#equal').simulate('click');

    expect(calc.default).toHaveBeenCalledWith('7*8');
    expect(comp.state('result')).toBe(56);
    expect(comp.find('#result').text().trim()).toBe('56');
  });

  it('should render exponentiation correctly', () => {
    const comp = shallow(<App />);

    comp.find('#d3').simulate('click');
    comp.find('#pow').simulate('click');
    comp.find('#d2').simulate('click');

    expect(comp.state('input')).toBe('3**2');

    comp.find('#equal').simulate('click');

    expect(calc.default).toHaveBeenCalledWith('3**2');
    expect(comp.state('result')).toBe(9);
    expect(comp.find('#result').text().trim()).toBe('9');
  });

  it('should evaluate the result on submit', () => {
    const comp = mount(<App />);    // mount is a must to simulate onSubmit

    comp.find('#d2').simulate('click');
    comp.find('#pow').simulate('click');
    comp.find('#d6').simulate('click');

    expect(comp.state('input')).toBe('2**6');

    comp.find('form').first().simulate('submit');

    expect(calc.default).toHaveBeenCalledWith('2**6');
    expect(comp.state('result')).toBe(64);
    expect(comp.find('#result').text().trim()).toBe('64');
  });
});

