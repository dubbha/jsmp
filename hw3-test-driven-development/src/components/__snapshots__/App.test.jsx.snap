// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly on startup 1`] = `
<div>
  <div>
    <form
      autoComplete="off"
      className="calc__form"
      onSubmit={[Function]}
    >
      <input
        className="calc__input"
        id="input"
        onChange={[Function]}
        type="text"
        value="5*5"
      />
      <button
        className="calc__button calc__button-equal"
        id="equal"
        onClick={[Function]}
      >
         = 
      </button>
    </form>
    <span
      className="calc__result"
      id="result"
    >
       
      25
       
    </span>
  </div>
  <div>
    <button
      className="calc__button"
      id="d7"
      onClick={[Function]}
    >
       7 
    </button>
    <button
      className="calc__button"
      id="d8"
      onClick={[Function]}
    >
       8 
    </button>
    <button
      className="calc__button"
      id="d9"
      onClick={[Function]}
    >
       9 
    </button>
    <button
      className="calc__button"
      id="div"
      onClick={[Function]}
    >
       ÷ 
    </button>
  </div>
  <div>
    <button
      className="calc__button"
      id="d4"
      onClick={[Function]}
    >
       4 
    </button>
    <button
      className="calc__button"
      id="d5"
      onClick={[Function]}
    >
       5 
    </button>
    <button
      className="calc__button"
      id="d6"
      onClick={[Function]}
    >
       6 
    </button>
    <button
      className="calc__button"
      id="mult"
      onClick={[Function]}
    >
       × 
    </button>
  </div>
  <div>
    <button
      className="calc__button"
      id="d1"
      onClick={[Function]}
    >
       1 
    </button>
    <button
      className="calc__button"
      id="d2"
      onClick={[Function]}
    >
       2 
    </button>
    <button
      className="calc__button"
      id="d3"
      onClick={[Function]}
    >
       3 
    </button>
    <button
      className="calc__button"
      id="add"
      onClick={[Function]}
    >
       + 
    </button>
  </div>
  <div>
    <button
      className="calc__button"
      id="d0"
      onClick={[Function]}
    >
       0 
    </button>
    <button
      className="calc__button calc__button-clear"
      id="clear"
      onClick={[Function]}
    >
       C 
    </button>
    <button
      className="calc__button"
      id="pow"
      onClick={[Function]}
    >
       ^ 
    </button>
    <button
      className="calc__button"
      id="sub"
      onClick={[Function]}
    >
       − 
    </button>
  </div>
</div>
`;
